SocketRocket/SRWebSocket.m

```diff
@@ -566,14 +602,31 @@ - (void)_failWithError:(NSError *)error;
     });
 }
 
-- (void)_writeData:(NSData *)data;
+- (void)_writeData:(NSData *)data
 {
-    [self assertOnWorkQueue];
+    [self _writeData:data completion:nil];
+}
 
+- (void)_writeData:(NSData *)data completion:(nullable SRSendCompletionBlock)completion
+{
+    [self assertOnWorkQueue];
+    
     if (_closeWhenFinishedWriting) {
+        if (completion) {
+            completion(SRErrorWithCodeDescription(2134, @"socket is closed"));
+        }
         return;
     }
-
+    
+    if (completion != nil) {
+        NSUInteger location = dispatch_data_get_size(_outputBuffer);
+        NSRange dataRange = NSMakeRange(location, data.length);
+        SRDataCallback *record =  [[SRDataCallback alloc] initWithRange:dataRange completion:completion];
+
+        static NSUInteger keyCount = 0;
+        _sendCallbacks[@(keyCount++)] = record;
```

Comment:
Not sure if this is the best way to use a `key` here.
Can we use a `[NSValue valueWithRange:dataRange]` as a key?